reps/2022-09-19-ray-on-spark.md

```diff
@@ -0,0 +1,258 @@
+Ray node requires listening on several port, a spark cluster might be shared by many users,
+each users might setup their own ray cluster concurrently, to avoid port conflicts,
+we can randomly select free port and start Ray node service,
+if failed we can retry on another free port.
+
+
+#### How much memory shall we allocate to Ray node service ( set via ray script --memory option) ?
+Spark does not provide explicit API for getting task allowed memory,
+SPARK_WORKER_MEMORY / SPARK_WORKER_CORES * RAY_NODE_NUM_CPUS
+
+
+#### How will we allocate memory for the Ray object store?
+We will calculate the memory reserved for the Ray object store in each Ray node as follows:
+
+1. Calculate the available space in the ``/dev/shm`` mount.
+2. Divide the ``/dev/shm`` available space by the number of Spark tasks (Ray nodes) that can run
+   concurrently on a single Spark worker node, obtaining an upper bound on the per-Ray-node object
+   store memory allocation.
+3. To provide a buffer and guard against accidental overutilization of mount space, multiply the
+   upper bound from (2) by ``0.8``.
+
+This allocation will ensure that we do not exhaust ``/dev/shm`` storage and accidentally spill to
+disk, which would hurt performance substantially.
+
+
+#### How to make ray respect spark GPU resource scheduling ?
+In spark task, we can get GPU IDs allocated to this task, so, when launching
+Ray node, besides specifying `--num-gpus` options, we need specify `CUDA_VISIBLE_DEVICES`
+environment so that we can restrict Ray node only uses the GPUs allocated to corresponding spark tasks.
+
+
+#### How to support all options of ray start
+Provide a ray_node_options argument (dict type).
+
+
+#### Where should the code live: ray repo or spark repo?
+Ray repo.
+Putting in spark repo is hard, It would be very hard to pass vote.
+Past examples includes horovod, petastorm, xgboost, tensorflow-spark, all of them tried to put in spark repo but failed.
+
+
+#### Does it support autoscaling ?
+Currently no. This implementation relies on spark barrier mode job, which does not support autoscaling currently.
+
+
+#### What’s the level of isolation spark provides ?
+Barrier mode spark task provides process-level isolation.
+No VM/container level isolation. Spark task does not support it.
+
+
+#### Custom resources support
+TPUs, specific accelerators, etc.
+Uncommon scenario, We can do it later?. Spark doesn't support this, spark can only schedule CPU, GPU, memory.
+
+
+### API Proposal
+
+#### Initialize ray cluster on spark API
+
+```
+ray_cluster_on_spark = ray.spark.init_cluster(num_cpus, num_gpus, memory)
+```
+
+Or
+
+```
+ray_cluster_on_spark = ray.spark.init_cluster(num_spark_tasks)
+```
+
+Initialize a ray cluster on the spark cluster, the arguments specified the ray cluster can use how much cpus / gpus / memory.
+And connect to the cluster.
+
+Returns an instance of type `RayClusterOnSpark``
+
+
+For each ray node, it might have different / random CPUs assigned ,
```

Comment:
This is not true any more: the spark task resource assignment is fixed.

Reply:
Definitely. Will update once I get authorized access to the branch.